go/vt/vtgate/executor.go

```diff
@@ -159,6 +159,17 @@ func (e *Executor) execute(ctx context.Context, safeSession *SafeSession, sql st
 	stmtType := sqlparser.Preview(sql)
 	logStats.StmtType = sqlparser.StmtType(stmtType)
 
+	// Mysql warnings are scoped to the current session, but are
+	// cleared when a "non-diagnostic statement" is executed:
+	// https://dev.mysql.com/doc/refman/8.0/en/show-warnings.html
+	//
+	// To emulate this behavior, clear warnings from the session
+	// for all statements _except_ SHOW, so that SHOW WARNINGS
+	// can actually return them.
+	if stmtType != sqlparser.StmtShow {
+		safeSession.Warnings = nil
```

Comment:
It just occurred to me that this assignment can race with itself because of recursive calls through engine and vindex. It will be safer to mutex this.

Reply:
Good point.

But it’s not simply solved with a muted. It also seems like the recursive vindex execute could clear warnings from the current session.

In the current use case I don’t think this is a problem because warnings get applied at the end of the route phase which comes after the vindex statements.

But this also means the partial scatter option doesn’t get propagated to the vindex statements themselves which it probably should.

That’s a long way of saying that this feature is still not totally defined for lookup-dependent plans.

IMO we should add the mutex to be race free and merge as is. Then we can fix this later.

Reply:
Sounds good.